File: pdb2fasta.cpp
#include<bits/stdc++.h>
#include <iostream>
using namespace std;

int main()
{

    freopen ("out.fa","w",stdout);
    fstream in;
    string strFileName = "in.pdb";
	in.open(strFileName.c_str());
	string sa,sb,sc;
        sa="";
        sb="";
        sc="";
    while( !in.eof()) //如果没有读到文件结尾
    {
        string strBuff;
        getline(in,strBuff);  //strBuff即 一行的数据

        istringstream is(strBuff);
        string s;
        int num =0;
        is>>s;
        if (s!="ATOM") continue;


        while(is>>s)
        {
            num++;
            if (num==3) sa=s;

            //if (num ==4 ) cout<<"\nn="<<s<<endl;
            if (num == 5 && sb==s) break;
            else if (num == 5 )
            {
                sb=s;
                s=sa;

                if (s=="ALA") cout<<"A";
                else if (s=="CYS") cout<<"C";
                else if (s=="ASP") cout<<"D";
                else if (s=="GLU") cout<<"E";
                else if (s=="PHE") cout<<"F";
                else if (s=="GLY") cout<<"G";
                else if (s=="HIS") cout<<"H";
                else if (s=="ILE") cout<<"I";
                else if (s=="LYS") cout<<"K";
                else if (s=="LEU") cout<<"L";
                else if (s=="MET") cout<<"M";
                else if (s=="ASN") cout<<"N";
                else if (s=="PRO") cout<<"P";
                else if (s=="GLN") cout<<"Q";
                else if (s=="ARG") cout<<"R";
                else if (s=="SER") cout<<"S";
                else if (s=="THR") cout<<"T";
                else if (s=="VAL") cout<<"V";
                else if (s=="TRP") cout<<"W";
                else if (s=="TYR") cout<<"Y";
                //else cout <<"  "<<s<<"   ";
            }

        }
        //cout<<endl;
        //cout<<endl;
    }
    in.close();

    return 0;
}

File: pdb2fasta.py
import sys

if len(sys.argv) <= 1:
    print 'usage: python pdb2fasta.py file.pdb > file.fasta'
    exit()
    
input_file = open(sys.argv[1])

letters = {'ALA':'A','ARG':'R','ASN':'N','ASP':'D','CYS':'C','GLU':'E','GLN':'Q','GLY':'G','HIS':'H',
           'ILE':'I','LEU':'L','LYS':'K','MET':'M','PHE':'F','PRO':'P','SER':'S','THR':'T','TRP':'W',
           'TYR':'Y','VAL':'V'}
print '>',sys.argv[1]
prev = '-1'
for line in input_file:
    toks = line.split()
    if len(toks)<1: continue
    if toks[0] != 'ATOM': continue
    if toks[5] != prev:
        sys.stdout.write('%c' % letters[toks[3]])
    prev = toks[5]

sys.stdout.write('\n')
input_file.close()
